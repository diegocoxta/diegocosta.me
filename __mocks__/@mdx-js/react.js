module.exports = {
  MDXProvider: jest.fn(({ children }) => children),
};
