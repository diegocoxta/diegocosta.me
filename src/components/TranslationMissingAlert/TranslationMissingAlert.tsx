import React from 'react';
import styled from 'styled-components';

import { usei18n, Link } from '~/utils/i18n';
import FixedContainer from '~/components/FixedContainer';

const Content = styled.div`
  background: ${({ theme }) => theme.accentColor};
  padding: 20px;
  color: ${({ theme }) => theme.backgroundColor};
  margin: 20px 0;
`;

const Paragraph = styled.p`
  margin: 5px 0;
`;

const CustomLink = styled(Link)`
  color: #ffffff;
  text-decoration: none;
  border-bottom: 1px dotted #ffffff;
`;

interface TranslationMissingAlertProps {
  slug: string;
  pageLanguage: string;
  translations?: (string | null | undefined)[] | null | undefined;
}

const renderSeparator = (index: number, length: number, separator: string): string => {
  if (index === 0 || index >= length) {
    return '';
  }

  return index + 1 !== length ? ', ' : ` ${separator} `;
};

export default function TranslationMissingAlert(props: TranslationMissingAlertProps): React.ReactElement {
  if (props.translations?.includes(props.pageLanguage)) {
    return <></>;
  }

  const i18n = usei18n();
  const title = i18n.getTranslationFor('translationMissingAlert.title');
  const subtitle = i18n.getTranslationFor('translationMissingAlert.subtitle');
  const separator = i18n.getTranslationFor('translationMissingAlert.separator');

  return (
    <FixedContainer>
      <Content data-testid="translation-missing-alert">
        <Paragraph>{title}</Paragraph>
        <Paragraph data-testid="translation-missing-alert-subtitle">
          {subtitle}{' '}
          {props.translations?.map((lang, index) => (
            <span key={`translation-missing-alert-${lang}`}>
              {renderSeparator(index, props.translations?.length ?? 0, separator)}
              <CustomLink language={lang} to={props.slug}>
                {i18n.getTranslationFor(`languages.${lang}`)}
              </CustomLink>
            </span>
          ))}
          .
        </Paragraph>
      </Content>
    </FixedContainer>
  );
}

TranslationMissingAlert.renderSeparator = renderSeparator;
