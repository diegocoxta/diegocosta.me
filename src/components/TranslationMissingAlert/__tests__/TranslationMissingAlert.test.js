import React from 'react';
import { render } from '@testing-library/react';

import TranslationMissingAlert from '../TranslationMissingAlert';

describe('<TranslationMissingAlert />', () => {
  it('renders properly with 2 languages available', () => {
    const { baseElement, getByText, queryByTestId, queryByText } = render(
      <TranslationMissingAlert slug="/path-to-article" pageLanguage="pt" translations={['en', 'es']} />
    );

    expect(queryByTestId('translation-missing-alert')).toBeTruthy();
    expect(getByText('languages.es')).toBeTruthy();
    expect(getByText('languages.en')).toBeTruthy();
    expect(queryByText('languages.pt')).toBeFalsy();
    expect(baseElement).toMatchSnapshot();
  });

  it('renders properly with 1 languages available', () => {
    const { baseElement, getByText, queryByTestId, queryByText } = render(
      <TranslationMissingAlert slug="/path-to-article" pageLanguage="pt" translations={['en']} />
    );

    expect(queryByTestId('translation-missing-alert')).toBeTruthy();
    expect(getByText('languages.en')).toBeTruthy();
    expect(queryByText('languages.es')).toBeFalsy();
    expect(queryByText('languages.pt')).toBeFalsy();
    expect(baseElement).toMatchSnapshot();
  });

  it('does not renders because the page exists', () => {
    const { baseElement, queryByText, queryByTestId } = render(
      <TranslationMissingAlert slug="/path-to-article" pageLanguage="pt" translations={['pt', 'en', 'pt']} />
    );

    expect(queryByTestId('translation-missing-alert')).toBeFalsy();
    expect(queryByText('languages.en')).toBeFalsy();
    expect(queryByText('languages.es')).toBeFalsy();
    expect(queryByText('languages.pt')).toBeFalsy();
    expect(baseElement).toMatchSnapshot();
  });

  it('renders the language string properly', () => {
    expect(TranslationMissingAlert.renderSeparator(0, 0, 'y')).toBe('');
    expect(TranslationMissingAlert.renderSeparator(0, 3, 'y')).toBe('');
    expect(TranslationMissingAlert.renderSeparator(1, 3, 'y')).toBe(', ');
    expect(TranslationMissingAlert.renderSeparator(2, 3, 'y')).toBe(' y ');
    expect(TranslationMissingAlert.renderSeparator(2, 3, 'e')).toBe(' e ');
    expect(TranslationMissingAlert.renderSeparator(2, 3, 'and')).toBe(' and ');
    expect(TranslationMissingAlert.renderSeparator(3, 3, 'y')).toBe('');
    expect(TranslationMissingAlert.renderSeparator(4, 3, 'y')).toBe('');
  });
});
