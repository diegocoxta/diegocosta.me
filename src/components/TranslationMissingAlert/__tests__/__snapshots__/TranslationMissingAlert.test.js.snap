// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TranslationMissingAlert /> does not renders because the page exists 1`] = `
<body>
  <div />
</body>
`;

exports[`<TranslationMissingAlert /> renders properly with 1 languages available 1`] = `
.c0 {
  max-width: 800px;
  margin: 0 auto;
  padding: 0 20px;
}

.c1 {
  padding: 20px;
  margin: 20px 0;
}

.c2 {
  margin: 5px 0;
}

.c3 {
  color: #ffffff;
  -webkit-text-decoration: none;
  text-decoration: none;
  border-bottom: 1px dotted #ffffff;
}

<body>
  <div>
    <section
      class="c0"
    >
      <main>
        <div
          class="c1"
          data-testid="translation-missing-alert"
        >
          <p
            class="c2"
          >
            translationMissingAlert.title
          </p>
          <p
            class="c2"
            data-testid="translation-missing-alert-subtitle"
          >
            translationMissingAlert.subtitle 
            <span>
              
              <a
                class="c3"
                href="/path-to-article"
                language="en"
              >
                languages.en
              </a>
            </span>
            .
          </p>
        </div>
      </main>
    </section>
  </div>
</body>
`;

exports[`<TranslationMissingAlert /> renders properly with 2 languages available 1`] = `
.c0 {
  max-width: 800px;
  margin: 0 auto;
  padding: 0 20px;
}

.c1 {
  padding: 20px;
  margin: 20px 0;
}

.c2 {
  margin: 5px 0;
}

.c3 {
  color: #ffffff;
  -webkit-text-decoration: none;
  text-decoration: none;
  border-bottom: 1px dotted #ffffff;
}

<body>
  <div>
    <section
      class="c0"
    >
      <main>
        <div
          class="c1"
          data-testid="translation-missing-alert"
        >
          <p
            class="c2"
          >
            translationMissingAlert.title
          </p>
          <p
            class="c2"
            data-testid="translation-missing-alert-subtitle"
          >
            translationMissingAlert.subtitle 
            <span>
              
              <a
                class="c3"
                href="/path-to-article"
                language="en"
              >
                languages.en
              </a>
            </span>
            <span>
               translationMissingAlert.separator 
              <a
                class="c3"
                href="/path-to-article"
                language="es"
              >
                languages.es
              </a>
            </span>
            .
          </p>
        </div>
      </main>
    </section>
  </div>
</body>
`;
